### PR TITLE
New article on gzipping AWS requests using boto3

### PR DESCRIPTION
**Pull request by lmammino**

### CODE REVIEW

File: content/posts/2021-06-20_how-to-send-gzipped-requests-with-boto3/index.md [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+
+## Send gzipped metrics to CloudWatch using boto3 
+
+Ok, this is the TLDR; a little gift for the ones in a rush that are looking for a quick _copy-pastable_ solution.
+
+Be aware that the various AWS SDKs are just a convenience layer in front of the AWS HTTP API. Python and `boto3` are no exception. Every time you are calling a method on a `boto3` client, you are just sending HTTP requests to AWS behind the scenes...
+
+This is how you can intercept and modify such HTTP requests before they are sent to AWS:
+
+  1. `boto3` has a built-in event system that, among other things, it allows you to intercept (and even modify) HTTP requests that are ready to be forwarded to AWS.
```
Comment (eoinsha) [inline content/posts/2021-06-20_how-to-send-gzipped-requests-with-boto3/index.md:39]: `, it allows` -> `, allows`. no need for that it, I think

File: content/posts/2021-06-20_how-to-send-gzipped-requests-with-boto3/index.md [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+Now, if you are curious to know more about my use case and how the `boto3` events system works you are more than welcome to keep reading the rest of the article.
+
+
+## The use case: sending custom metrics to CloudWatch
+
+Last week, during my work at [fourTheorem](https://www.fourtheorem.com/), we started to get intermittent alarms for a Lambda in our stack that was failing because of requests to CloudWatch being occasionally throttled.
+
+Once we looked at the error, the problem was actually quite obvious, but let me give you a simplified overview of what we are doing with this Lambda.
+
+This Lambda is triggered by a Kinesis stream in which we publish custom metric data from all the other components of our application. The Lambda reads these metrics and publishes them to CloudWatch using the [`PutMetricData` API](https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_PutMetricData.html).
```
Comment (eoinsha) [inline content/posts/2021-06-20_how-to-send-gzipped-requests-with-boto3/index.md:98]: This is all based on the ideas seen in Yan's async custom metrics stack (https://github.com/lumigo-io/SAR-async-lambda-metrics) btw

File: content/posts/2021-06-20_how-to-send-gzipped-requests-with-boto3/index.md [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+```
+
+More details can be found in the documentation for the [`MetricDatum` type](https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_MetricDatum.html).
+
+As you can tell from the comments in the snippet above, the API has some interesting limits to be considered:
+
+  - Up to **20** different metrics
+  - Up to **10** dimensions per metric
+  - Up to **150 values** per metric
+  - Up to **40 KB** in size for HTTP POST requests
```
Comment (eoinsha) [inline content/posts/2021-06-20_how-to-send-gzipped-requests-with-boto3/index.md:141]: It's really good that you are summarising all these limits here.

File: content/posts/2021-06-20_how-to-send-gzipped-requests-with-boto3/index.md [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+
+At this point, I thought _"Ok, probably `boto3` is automatically doing the compression for us, because why shouldn't it?"_
+
+But if you have used AWS for long enough, you learn not to give too many things for granted, so... Let's test this assumption, first!
+
+
+## Testing `boto3` default behavior
+
+So, how can we see if the by default `boto3` is already gzipping our requests?
+Requests are going to AWS and I could think of 3 different ways to inspect how the final HTTP request actually looks like when sent to AWS:
```
Comment (eoinsha) [inline content/posts/2021-06-20_how-to-send-gzipped-requests-with-boto3/index.md:161]: `like` is superfluous I think. `how it looks` makes sense on its own, or `what it looks like`

File: content/posts/2021-06-20_how-to-send-gzipped-requests-with-boto3/index.md [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+Compressing the payload for `PutMetricData` requests is not an absolute guarantee that we won't be exceeding the 40 KB payload limit. Gzip certainly helps, but you might still bump into the limit and have failures in your application.
+
+If you want to be _failure-proof_ ™️, here are some ideas that we explored with our colleagues:
+
+  - Optimistic solution: Try the request and catch potential exceptions. If you see the `PutMetricData` failed because the payload is too large, just split all the data into 2 requests.
+  - Pessimistic solution: pre-compute the gzipped payload before sending the request and if that's bigger than 40 KB, then split the request into 2 parts.
+
+The pessimistic solution has the disadvantage that you will be gzipping the payload twice (before you create the request to make sure it's within boundaries and then in your event handler), but it will avoid sending invalid requests upfront.
+
+Right now, we haven't implemented any of these strategies. In our case, these metrics are not business-critical and we can afford some sparse failure, so we preferred to keep the code as simple as possible.
```
Comment (eoinsha) [inline content/posts/2021-06-20_how-to-send-gzipped-requests-with-boto3/index.md:338]: Aside: I would question whether the metrics are not business critical in our case!!

File: content/posts/2021-06-20_how-to-send-gzipped-requests-with-boto3/index.md [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+If you found this article useful [consider following me on Twitter](https://twitter.com/loige) and feel more than welcome to leave a comment below. I'd be really curious to find out what was your use case and if this article helped you out.
+
+A huge "thank you" goes to my colleague Martin for involving me in this piece of work (and indirectly for dragging me into this rabbit hole 🐇)!
+
+See you soon! 👋
+
+
+### About fourTheorem
+
+If you want to know more about [fourTheorem](https://www.fourtheorem.com/), we are a team of business-focused technologists that deliver. We have been helping dozen of companies to get the best out of AWS and we had a lot of fun while doing that. If you are curious you can check out some of our [case studies](https://www.fourtheorem.com/case-studies) and our [customer success stories](https://www.fourtheorem.com/customers).
```
Comment (eoinsha) [inline content/posts/2021-06-20_how-to-send-gzipped-requests-with-boto3/index.md:360]: Cool! Maybe `a dozen companies` or `dozens of companies`?

**Review by eoinsha — COMMENTED** [review]
This is a really good article. I would be very appreciative if I had found this facing similar problems. I love the TL;DR followed by a super-detailed breakdown! 👏

File: content/posts/2021-06-20_how-to-send-gzipped-requests-with-boto3/index.md (hunk 2 above)
Comment (lmammino) [reply content/posts/2021-06-20_how-to-send-gzipped-requests-with-boto3/index.md:98]: Good point, it's definitely worth adding that!

**Pull request merged**